### PR TITLE
HZN-1420: Add support for indexing alarms in Elasticsearch.

### PR DESCRIPTION
**Pull request by j-white**
JIRA: https://issues.opennms.org/browse/HZN-1420

Here we add support for indexing alarms in Elasticsearch as an alternative (and eventual replacement) for the existing alarm-change-notifier plugin. This is accomplished by hooking into the AlarmLifecycleListener interface, and indexing the alarms in Elasticsearch as they change.

The alarms are indexed in a fashion that allows you to query the state of any or all alarms at some given point in time.

### CODE REVIEW

File: features/alarms/history/rest/api/src/main/java/org/opennms/features/alarms/history/rest/api/AlarmHistoryRestService.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ *
+ * /alarm/history    -> all as of now
+ * /alarm/history/9  -> specific alarm now
+ * /alarm/history/9/journal -> journal history of alarm #9
+ *
+ * ?match-type=alarm-id   (default to =reduction-key)
+ * ?at=3423842384834      (default to =$now)
+ */
+@Path("alarms/history")
```
Comment (j-white) [inline features/alarms/history/rest/api/src/main/java/org/opennms/features/alarms/history/rest/api/AlarmHistoryRestService.java]: This REST API endpoint should be documented in the develop guide.

**Comment by j-white** [comment]
The indexing logic is implemented in the ElasticAlarmIndexer class, and the querying (used for testing and validation) is implemented in the ElasticAlarmHistoryRepository class.

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](http://www.opennms.org/documentation/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit http://www.opennms.org/wiki/Contributor_Agreement

File: opennms-doc/guide-admin/src/asciidoc/text/alarms/history.adoc [hunk 2]
```diff
@@ -0,0 +1,137 @@
+[[ga-alarm-history]]
+=== Alarm History
+
+The _Alarm History_ feature integrates with _Elasticsearch_ to provide long term storage and maintain a history of alarm state changes.
+
+When enabled, alarms are indexed in _Elasticsearch_ when they are created, deleted, or when any of the "interesting" fields on the alarm are updated (more on this bellow.)
```
Comment (mattixtech) [inline opennms-doc/guide-admin/src/asciidoc/text/alarms/history.adoc]:
```suggestion
When enabled, alarms are indexed in _Elasticsearch_ when they are created, deleted, or when any of the "interesting" fields on the alarm are updated (more on this below.)
```

**Comment by opennms-bamboo** [comment]
Welcome Jesse White (jesse@opennms.com).

We would like to thank you for your contribution.

I could not find a signed OpenNMS Contributor Agreement (OCA) assigned to the username Jesse White (jesse@opennms.com).
In order to maintain some order with respect to the copyright to the code,
every contributor is asked to sign a "Contributor Agreement".

The OpenNMS Contributor Agreement can be downloaded [here](http://www.opennms.org/documentation/ContributorAgreement.pdf).
Please send a signed copy to oca@opennms.org.

If you need more information, please visit http://www.opennms.org/wiki/Contributor_Agreement

**Comment by j-white** [comment]
Given that this is all new code and the feature is optional, I'm going to go ahead and merge this.

If there's any feedback after the fact, feel free to comment on this PR, or contact me (`jesse`) in [chat](https://chat.opennms.com/opennms/channels/opennms-development).

**Pull request merged**